passes/hierarchy/hierarchy.cc

```diff
@@ -710,6 +711,411 @@ RTLIL::Wire *find_implicit_port_wire(Module *module, Cell *cell, const std::stri
+// at the counts dictionary. If the count is at most 1, we can still operate in
+// place: the module either only appears at top-level or only appears as the
+// instance into which we're binding. If the count is 2 or greater, we need to
+// take a copy. In this case, we take the copy and modify it. We then decrement
+// the count for the original and give the new copy a count of 1.
+//
+// Finally, this function updates top_mod if *top_mod is non-null and **top_mod
+// is a module that was updated.
+static bool
+insert_binding(RTLIL::Design               &design,
+               dict<RTLIL::IdString, int>  &counts,
+               RTLIL::Module               *mod,
+               const RTLIL::Binding        &binding,
+               RTLIL::Module              **top_mod)
+{
+	std::string desc = binding.describe();
+
+	// If we're not at top-level, look for a cell in mod with the right
+	// (hierarchical) name. This is what section 23.3.1 of IEEE 1800 refers
+	// to as "the local A.B.C".
+	if (mod) {
+		RTLIL::Binding::Path path = binding.find_rel_cell(design, *mod);
+		if (path.size()) {
+			log("Applying %s at a local path.\n", desc.c_str());
+			return bind_path(design, counts, path, binding, top_mod);
+		}
+	}
+
+	// If no luck, try to interpret the name as a hierarchical reference
+	// starting at the top level.
+	std::vector<RTLIL::Binding::Path> paths = binding.find_tl_cells(design);
+	if (!paths.empty()) {
+		bool did_something = false;
+		log("Applying %s at %d top-level path%s.\n",
+		    desc.c_str(), GetSize(paths), paths.size() > 1 ? "s" : "");
+		for (const auto &path : paths)
+			did_something |=
+				bind_path(design, counts, path, binding, top_mod);
+		return did_something;
+	}
+
+	// If neither of these worked, search for a module by name. Note that
+	// this takes no account of design hierarchy.
+	std::vector<RTLIL::Module *> mods = binding.find_concrete_module_targets(design);
+	if (!mods.empty()) {
+		// We can add the binding directly to the module, operating in place.
+		bool did_something = false;
+		log("Applying %s to %d module%s.\n",
+		    desc.c_str(), GetSize(mods), mods.size() > 1 ? "s" : "");
+		for (auto mod : mods)
+			did_something |= binding.bind_into(design, *mod);
+		return did_something;
+	}
+
+	log_error("Couldn't find a target for %s.\n", desc.c_str());
+}
+
+// Worker function for get_cell_type_counts. To avoid double counting, we
+// assume that any module that has already been processed has an entry in
+// counts. Top-level modules get a count of zero added just before calling this
+// function on them.
+static void
+get_cell_type_counts_worker(RTLIL::Design &design,
+                            dict<RTLIL::IdString, int> &counts,
+                            RTLIL::Module &mod)
+{
+	for (auto cell : mod.cells()) {
+		// Set counts[cell->type] to 1 if it isn't defined
+		// yet, or increment if it is.
+		auto pr = counts.insert(std::make_pair(cell->type, 1));
+		if (!pr.second)
+			++pr.first->second;
```

Comment:
It seems like this is equivalent to simply `counts[cell->type]++`, given the expected behavior of `operator[]` when not present in the `dict`.

Reply:
If the key isn't present in the dictionary, I think the value gets default-constructed. But I don't think there's any guarantee that an `int` will be constructed as zero is there?

Reply:
Given the general requirement that map values should be default constructible, I would think it would be initialized to zero. I think I have seen this behavior relied upon in the past, but I could be wrong. This is what `dict` uses in the not-present case:
```C++
i = do_insert(std::pair<K, T>(key, T()), hash);
```

While I'm not opposed to doing a clearer, unobjectionable thing, I am very interested in learning what actually happens here.

Reply:
> If the key isn't present in the dictionary, I think the value gets default-constructed. But I don't think there's any guarantee that an `int` will be constructed as zero is there?

Yes, the in default constructor sets it to zero.

However, notice that you have to run in. If you have a class member that is an int, then it's only default constructed when the constructor is called, otherwise it is left uninitialized:
https://godbolt.org/z/zPPYP5dTb

In the case of a `dict<T, int>` (or `unordered_map<T, int>` and other std containers) the values are always default constructed, i.e. ints are set to zero.

Reply:
Oh, neat! I've just learned something, thanks :-)

Reply:
Incidentally, the code just below this note was looking at `pr.second` to see whether we'd seen a match, which isn't possible any more. The obvious translation looks like it might do a second lookup and (after compiling and looking at the generated binary) it seems that it does :-(

The result is simpler and more clearly right, so I'll switch anyway but it's a pity that GCC (10.3.8) can't fold this away.